src/backend/distributed/commands/citus_add_local_table_to_metadata.c

```diff
@@ -1158,27 +1158,62 @@ GetRenameStatsCommandList(List *statsOidList, uint64 shardId)
 
 
 /*
- * DropDefaultExpressionsAndMoveOwnedSequenceOwnerships drops default column
- * definitions for relation with sourceRelationId. Also, for each column that
- * defaults to an owned sequence, it grants ownership to the same named column
- * of the relation with targetRelationId.
+ * DropNextValExprsAndMoveOwnedSeqOwnerships drops default column definitions
+ * that are based on sequences for relation with sourceRelationId.
+ *
+ * Also, for each such column that owns a sequence, it grants ownership to the
+ * same named column of the relation with targetRelationId.
  */
 static void
-DropDefaultExpressionsAndMoveOwnedSequenceOwnerships(Oid sourceRelationId,
-													 Oid targetRelationId)
+DropNextValExprsAndMoveOwnedSeqOwnerships(Oid sourceRelationId,
+										  Oid targetRelationId)
 {
 	List *columnNameList = NIL;
 	List *ownedSequenceIdList = NIL;
 	ExtractDefaultColumnsAndOwnedSequences(sourceRelationId, &columnNameList,
 										   &ownedSequenceIdList);
 
+	uint16 sourceRelDefExprArrIndex = 0;
+
 	char *columnName = NULL;
 	Oid ownedSequenceId = InvalidOid;
 	forboth_ptr_oid(columnName, columnNameList, ownedSequenceId, ownedSequenceIdList)
 	{
-		DropDefaultColumnDefinition(sourceRelationId, columnName);
+		HeapTuple columnTuple = SearchSysCacheAttName(sourceRelationId, columnName);
+		if (!HeapTupleIsValid(columnTuple))
+		{
+			ereport(ERROR, (errcode(ERRCODE_UNDEFINED_COLUMN),
+							errmsg("column \"%s\" of relation \"%s\" does not exist",
+								   columnName, get_rel_name(sourceRelationId))));
+		}
+
+		Form_pg_attribute columnForm = (Form_pg_attribute) GETSTRUCT(columnTuple);
+		bool columnHasDefaultExpr = columnForm->atthasdef;
+
+		ReleaseSysCache(columnTuple);
 
-		/* column might not own a sequence */
+		if (columnHasDefaultExpr)
+		{
+			if (DefExprContainsNextVal(sourceRelationId, sourceRelDefExprArrIndex))
```

Comment:
are we sure that using an `index` is good enough? What happens if a column is dropped? Shouldn't it be safer if we use `attrNo` of the column instead?

Reply:
Unlike `TupleDesc` array, dropped attributes are deleted from `defval` and `defval` array stores default expressions in the order they appear in `TupleDesc` array.

Also, the way we access `defval` array is pretty much similar to what we do in `citus_ruleutils.c/pg_get_tableschemadef_string()`.

However, if you would feel more comfortable, then I can change the function so that it searches for the default expression based on attrNos stored in `defval[i].adnum` (AttrDefault.adnum).

Reply:
done